### DISCUSSION TITLE
Two Equation Turbulence Model

### DISCUSSION THREAD

**Discussion by suqingji** [idaholab/moose#20854] [Q&A Modules: Navier-Stokes] [answered]
As mentioned elsewhere, the finite element module of MOOSE does not have a two-equation turbulence model. Now we are trying to add this module, but we have encountered many problems. I wonder whether any experts have tried to add a two-equation turbulence model, or which users have done such work based on MOOSE.
Thanks

**Comment by GiudGiud** [comment, marked as answer]
@lindsayad

**Reply by suqingji** [reply]
In fact, I think I saw an expert who developed the K-Epsilon turbulence model in 2017, but it seems to have broken down halfway through, and now I'm developing the K-Omega turbulence model, which is able to do some simple examples, and the results look good, but in a slightly more complicated problem, there are convergence problems, I wonder if there is an opportunity to exchange such questions with experts in this question.

**Reply by lindsayad** [reply]
There is an issue for this: https://github.com/idaholab/moose/issues/9007. From that issue you can link to some commits that I worked on a very long time ago. We (the MOOSE team) do not have programmatic funding to work on the finite element side of the `navier_stokes` module, but we would love any outside contribution that added that capability. I do not know whether my old commits would help you in any way, but please feel free to work from them if they are of any help. If you do decide to develop this yourself, I may be able to provide some limited coaching as my time allows

**Reply by suqingji** [reply]
Hi, Prof. lindsayad
Thanks for your reply. I would like to describe some of the problems I encountered in the development of turbulence programs.
I had finished a kind of two turbulence model which is Wilcox k-omega turbulence model, and I had finished testing some important cases like the "2D Zero Pressure Gradient Flat Plate Verification Case" and “backward-facing-step Case". When I simulated the above simple cases, convergence is fine and from the computed solution I could find that the relationship between u+ and log(y+) compared with the DNS solution is about the same in the "2D Zero Pressure Gradient Flat Plate Verification Case". But in other complex cases, I always face a big issue which is a convergence problem. As I know, in other FVM programs we need to use some method to stabilize the simulation for turbulence equation in particular the production and dissipation term, but in FEM programs I wonder if there is a method to stabilize simulation like SUPG/PSPG.
Looking forward to your reply. Thanks.

**Reply by lindsayad** [reply]
In the `navier_stokes` module in our FEM implementation, we have both SUPG and PSPG available. However, if you want to add a turbulence model, you would have to implement the SUPG stabilization of the additional equations yourself, e.g. you would have to develop the SUPG stabilization of k and epsilon for a k-epsilon model. Additionally if you add an additional term like a Reynolds stress for the turbulent momentum transport, you would have to add that term into the existing SUPG and PSPG stabilization of the momentum and mass equations.

**Reply by lindsayad** [reply]
If you have not seen it already, we have an article on the FEM portion of the `navier_stokes` module. The journal article is [here](https://www.sciencedirect.com/science/article/abs/pii/S0965997817310591); if you do not have access, there is an arXiv version of it [here](https://arxiv.org/pdf/1710.08898.pdf).

**Reply by suqingji** [reply]
Thank you for your reply in spite of your busy schedule. I am sorry that my description may not be accurate, so you did not understand what I mean.
Yes, I did add SUPG stabilization to the k and omega equation according to the journal article named "Overview of the Incompressible Navier Stokes simulation capabilities in the MOOSE Framework". But due to the production terms of those turbulence equations are too complex to add the jacobian term to matrix by hand. Then I tried to add those jacobian terms by AD, which will cause non-convergence. In other words, my Jacobian matrix is right made by hand, but it is the production term of those turbulence equations is complex that causes the convergence problem even though I had added SUPG to those equations. So I wonder if there are need to add other stabilization methods to those turbulence equations to stabilize those turbulence equations' simulation, or if there are some other details worth noting I did not know.

**Reply by lindsayad** [reply]
> Then I tried to add those jacobian terms by AD, which will cause non-convergence.

I doubt that the AD-generated Jacobian caused non-convergence. You can check the accuracy of your Jacobian by running with the PETSc command-line option `-snes_test_jacobian`.

What Reynolds numbers are you trying to run at? Maybe @tanoret would have some comment on stability of the k-epsilon equations

**Reply by suqingji** [reply]
Thanks for your reply,
Your advice is of great help to me. I had finished running with the PETSc command-line option '-snes_test_jacobian'. And I found that the Jacobian is not very accurate. However, that's not the root cause. I tried to run with the FDP(Finite Difference Preconditioner) which could build the perfect approximate Jacobian. But some otherwise solvable problems do not converge. So I doubt that the residual term of production and dissipation caused non-convergence. Maybe those terms need to be restricted, but I have never found references such as literature or code implementing restrictions by the finite element method.
Looking forward to your reply. Thanks.

**Reply by lindsayad** [reply]
I could help try to debug this, but you would need to submit a pull request for me to do so

**Reply by suqingji** [reply]
Thanks again for your reply, and I really appreciate you would like to help me to debug this by taking time out of your busy schedule.
Now I have finished pushing those codes to Github, you could pull it through `https://github.com/suqingji/i_flow.git`.
I have added a `readme.md` file to `/i_flow/case/backstep_sparse/`, if you have any questions about `i_flow`, please contact me at suqj5@mail2.sysu.edu.cn or here.

**Reply by lindsayad** [reply]
I’m still planning to look into this at some point. I’m so sorry I haven’t gotten to it yet. If we had RANS in FV would you use it, or are you sold on using FE? @tanoret is working on FV RANS and will be submitting a PR soon. I ran across your GitHub discussion https://github.com/mfem/mfem/discussions/3165 … definitely makes sense that you are checking some other options given how incredibly slow I’ve been 😊

**Reply by suqingji** [reply]
Hi, @lindsayad

Here I come out with a new problem about my FEM RANS codes. #https://github.com/idaholab/moose/discussions/21996

In a word, my code can be used to simulate some simple cases, but it becomes non-converged in a complex problem. Besides, I have found some bugs in the Navier-Stokes module.


Best regards.

**Reply by lindsayad** [reply]
We can continue to discuss the specific issue in #21996 over in #21996. Regarding your work on your two equation turbulence model, it sounds like you are making some progress. Do you think you would ever contribute your k-epsilon FE implementation to MOOSE?

**Reply by suqingji** [reply]
Dear @lindsayad,

After finishing my k-omega model, I will do the k-epsilon RANS FE implementation on MOOSE. I think it is my pleasure to contribute my k-epsilon FE implementation to MOOSE. 


QINGJI SU

**Reply by lindsayad** [reply]
Awesome. k-omega would be a welcome contribution as well :smile:

**Reply by suqingji** [reply]
No problem

**Reply by suqingji** [reply]
Hi, @lindsayad 

As I mentioned above. I had finished a Wilcox k-omega even though which just can be used to simulate some simple cases.
Now, I have finished part function of some k-omega models such as Wilcox k-omega, SST, and BSL versions. And I have finished some test cases. Unfortunately, They just can simulate some simple cases, too. And the solution is right.

Now, I am adding some k-epsilon models and a SA model to MOOSE.

When I was adding those RANS models to MOOSE, I found some important points which are we must care about in implementing the RANS model.

Here are:

1. In complex problems, we need separate the solver for many reasons. For example, the matrix is too big to solve.
2. We need `Crosswind Diffusion Stabilization` (CD Stabilization) like `SUPG` to stabilize the compute progress.
3. Mabey we also need some limiter and Flux Correction.

About the first point, I had known that MOOSE teams plan to add separate solvers.
About the second point, I am adding `CD Stabilization` to MOOSE framework base kernels. But I have faced many problems,. So could you please help me add `CD Stabilization` to MOOSE in your applicable time?
About the third point, I am studying those problems.

When we finished the `CD Stabilization` and Automatic boundary on turbulence equation. I think it is a suitable time to contribute my codes to MOOSE because the turbulence function is thorough.

Thank you and best regards.

**Reply by lindsayad** [reply]
It sounds like you have made some excellent progress!

> So could you please help me add CD Stabilization to MOOSE in your applicable time?

I would love to help. However, my ability to support will be a function of whether this fits into my funding program's needs. We have a common TH layer milestone this year which this *might* fit into. @lingzou do you think that crosswind diffusion stabilization would help finite element models in SAM?

> 3\. Mabey we also need some limiter and Flux Correction.

We have scalar limiting capability that we use in our finite volume discretizations. Some documentation can be found [here](https://mooseframework.inl.gov/syntax/Limiters/index.html). @grmnptr is planning to add vector limiters sometime this year. These are applicable when you want to limit the velocity (or other Euclidean vectors) and are generally more stable (e.g. more diffusive) than limiting each velocity component individually

**Reply by lingzou** [reply]
@lindsayad Thanks for including me in the discussion.

I have to be honest that I am not sure how useful the crosswind diffusion stabilization (CDS) will be to SAM applications. From some quick read of

https://doc.comsol.com/5.5/doc/com.comsol.help.comsol/comsol_ref_fluidflow.20.14.html
and the reference
G. Hauke and T.J.R. Hughes, “A Unified Approach to Compressible and Incompressible Flows”, Comp. Meth. Appl. Mech. Engrg, vol. 113, pp. 389–395, 1994.

it seems to me that the main purpose of CDS for shock-capturing operator, which I assumes the targeting application is compressible flows. The comsol website also states that, for incompressible flow, it won't capture a shock but still useful as it adds extra diffusion, so I suppose it will be useful.

The current SAM one-dimensional incompressible flow uses the combination of PSPG/SUPG stabilization scheme. Although not a perfect monotonicity preserving scheme as in some FVM scheme, it does a good job. So, from this perspective, I am not sure how useful it would be to add some extra diffusion. It might be more helpful towards 3d flow? From the reading of the reference paper, it appears to me that the CDS scheme requires certain access to local grid info (i.e., the g^{ij} term), which I believe libMesh/MOOSE should provide, so probably quite do-able.

**Reply by lindsayad** [reply]
Thanks for the detailed reply @lingzou. Based on that, @suqingji, I doubt that I will be able to work on crosswind diffusion with programmatic funding. It is possible that I might do it in free time, but it's probably not likely unfortunately

**Reply by suqingji** [reply]
Hi @lindsayad,
First, thanks for the detailed reply @lingzou @lindsayad.
As @lingzou mentioned above,  the main purpose of CDS is shock-capturing. In fact, CDS is also essential for incompressible flow. Because it can stabilize the simulation for some problems in which the variable has big shocks such as turbulence equation, of course, in incompressible flow without turbulence model it is also can help to stabilize the simulation, especially in 3D simulation.

In any way, thank you reply, I will try to add CDS to MOOSE.

By the way, I have another question about MOOSE, maybe which is a bug, I will @you under the question I raised earlier. would you mind answering the question in your spare time?

**Reply by lindsayad** [reply]
I'm asssuming you're talking about #21996? Yes happy to discuss anytime
